### DISCUSSION TITLE
How to install from source ("develop")?

### DISCUSSION THREAD

**Discussion by orbeckst** [MDAnalysis/mdanalysis#4636] [Installation]
What's the best way to install MDAnalysis from source, in particular the develop branch after you [forked the mdanalysis repo](https://userguide.mdanalysis.org/stable/contributing_code.html#forking-code-repo):
```bash
git clone https://github.com/your-user-name/mdanalysis.git
cd mdanalysis
git remote add upstream https://github.com/MDAnalysis/mdanalysis
```

The [installing develop instructions in the User Guide](https://userguide.mdanalysis.org/stable/contributing_code.html) are outdated (issue https://github.com/MDAnalysis/UserGuide/issues/368). What's the best/cleanest way to install?

1. using conda-forge packages (with `mamba`/`conda`)
2. using pip packages (`pip` and `virtualenv` ... and your compiler toolchain)

**Comment by orbeckst** [comment]
For a quick conda/mamba environment one can use
```bash
mamba create -n mda312dev python=3.12
mamba install -n mda312dev --only-deps mdanalysis mdanalysistests
mamba activate mda312dev
```
which makes use of `mamba --only-deps mdanalysis mdanalysistests` to pull the dependencies for the published mdanalysis and mdanalysistests conda-forge packages.

It has the disadvantage that these dependencies *may be different in the current source*.... but it's quick.

**Reply by IAlibay** [reply]
because of the disadvantage you list, doing this is a really bad idea - especially with the current state of the develop branch

**Comment by IAlibay** [comment]
Since we don't offer an environment file, the best approach is to just manually list all the deps to install imho.

We could offer an environment file, and I looked into this a while ago, but it requires an action to check our CI vs what we store.

Note: Before someone says "let's switch completely to environment files" - the source of truth is our CI deps build action, we don't use environment files because they are too inflexible, it really won't work, I tried.

**Reply by RMeli** [reply]
What about parsing the dips build action (it's a YAML file after all)? See https://github.com/MDAnalysis/UserGuide/issues/368#issuecomment-2234251162 for a quick stab at it.

**Reply by IAlibay** [reply]
If we want an env file then we can have one and create the necessary tooling to verify it (that's what I meant by checking what we store, it's fine to _me_, but it's extra tooling and every time I make extra tooling I get complaints about extra tooling).

**Reply by orbeckst** [reply]
Let's start simple with a working environment file, update the UG, and then see if we need CI to check that this env file agrees with what we use in CI.

(This was the conclusion from a discussion at the MDAKits meeting today.)

**Reply by orbeckst** [reply]
And once we have the env file in the rep, we can add a proper answer here (which will be mirrored in the UG.)

**Comment by marinegor** [comment]
@orbeckst I'm not sure if this was your initial request, but I think you can try following syntax of `pip` to install directly from git(hub):

```bash
# install default package
python -m pip install 'mdanalysis @ git+https://github.com/MDAnalysis/mdanalysis.git@develop#subdirectory=package'

# install a particular tag -- note the single quotes!
python -m pip install 'mdanalysis[analysis] @ git+https://github.com/MDAnalysis/mdanalysis.git@develop#subdirectory=package'
```

Note the single quotes in the latter option -- won't work without them (will install default branch). And this is for bash shell, not sure if it works for others without hard quoting (though I think you can safely quote the `git+...` part).

I usually do that in a virtual environment, which works perfectly fine, thanks to the dev team.

Also, for the sake of huge speed gain, I recommend doing this with [uv](https://github.com/astral-sh/uv), which supports exactly the same syntax (`uv pip install ...`), and also installs whole `[analysis]` tag in **1m30s** with a clean cache / **2s**  (`pip` installs [analysis] in 4 minutes with clean cache, and like 40 seconds with cached dependencies, i.e. 20 times slower).

**Reply by orbeckst** [reply]
That's a nice answer for pip installation without the cloned local repository.
